mllib/src/main/scala/org/apache/spark/ml/tree/impl/BaggedPoint.scala

```diff
@@ -33,13 +33,20 @@ import org.apache.spark.util.random.XORShiftRandom
  * this datum has 1 copy, 0 copies, and 4 copies in the 3 subsamples, respectively.
  *
  * @param datum  Data instance
- * @param subsampleWeights  Weight of this instance in each subsampled dataset.
- *
- * TODO: This does not currently support (Double) weighted instances.  Once MLlib has weighted
- *       dataset support, update.  (We store subsampleWeights as Double for this future extension.)
+ * @param subsampleCounts  Number of samples of this instance in each subsampled dataset.
+ * @param sampleWeight The weight of this instance.
  */
-private[spark] class BaggedPoint[Datum](val datum: Datum, val subsampleWeights: Array[Double])
-  extends Serializable
+private[spark] class BaggedPoint[Datum](
+    val datum: Datum,
+    val subsampleCounts: Array[Int],
+    val sampleWeight: Double) extends Serializable {
+
+  /**
+   * Subsample counts weighted by the sample weight.
+   */
+  def weightedCounts: Array[Double] = subsampleCounts.map(_ * sampleWeight)
```

Comment:
You can just store subsampleCounts * sampleWeights as a field, and then these two constructor params won't become fields. No need to compute it each time. On the other hand I don't mind this change as it reduces memory usage -- ints are smaller than doubles.

On the other other hand, if you often call this method with one sample, it could be more efficient, to even provide a second constructor taking an Int rather than array. Then you do need to store the array of doubles internally.

I don't feel strongly about it

Reply:
I was going to say I'm not sure as well - as you said there is a small trade off - also I would mention that for clarity in code I think it is a bit clearer what the field is this way as opposed to doing the computation beforehand.  However, I noticed this field isn't even used anywhere, it's just there for clarity, in here:

mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

we get a single subsample count and weight and they are passed separately to another method orderedBinSeqOp or mixedBinSeqOp

I guess it's nice to have this method for clarity or even debugging but if it's not used maybe I should just get rid of it altogether to limit confusion?  What are your thoughts on this?

Reply:
Yes I wouldn't carry it around if it can be avoided. How about a `def subsampleWeight(i: Int)` method instead, that computes it from subsampleCounts * sampleWeight?

Reply:
done! removed the method.  they are retrieved and passed down to several layers of methods before they get to the aggregator where that computation subsampleCount * sampleWeight happens, so it doesn't make sense to have any method there (unless it is just for some debugging or clarity in the code)